kuma/javascript/src/active-banner.jsx

```diff
@@ -182,86 +179,80 @@ function Banner(props: BannerProps) {
-        let bannersList = props.banners || defaultBanners;
-        for (const banner of bannersList) {
-            if (userData.waffle.flags[banner.id]) {
-                // if this banner is to be shown to authenticated users only,
-                // and the current user is not authenticated, return `null`
-                if (banner.authenticated && !userData.isAuthenticated) {
-                    return null;
-                }
+    const bannerId = userData.waffle.flags
+        ? BANNER_IDS.find(id => userData.waffle.flags[id] && !isEmbargoed(id))
+        : null;
 
-                // special case for the subscriptions banner as we need access
-                // to the username as it forms part of the `url`
-                if (banner.id === 'subscription_banner' && userData.username) {
-                    banner.url = `${window.mdn.wikiSiteUrl}/${locale}/profiles/${userData.username}/edit#subscription`;
-                }
+    if (!bannerId) {
+        return null;
+    }
+
+    switch (bannerId) {
+        case DEVELOPER_NEEDS_ID:
+            return (
+                <Banner
+                    id={bannerId}
+                    classname="developer-needs"
+                    title={gettext('MDN Survey')}
+                    copy={gettext(
+                        'Help us understand the top 10 needs of Web developers and designers.'
+                    )}
+                    cta={gettext('Take the survey')}
+                    url={
+                        'https://qsurvey.mozilla.com/s3/Developer-Needs-Assessment-2019'
+                    }
+                    newWindow
+                />
+            );
 
-                if (!isEmbargoed(banner.id)) {
-                    return <Banner {...banner} />;
-                }
+        case SUBSCRIPTION_ID:
+            if (!userData.isAuthenticated) {
+                return null;
             }
-        }
+
+            return (
+                <Banner
+                    id={bannerId}
+                    classname="mdn-subscriptions"
+                    title={gettext('Become a monthly supporter')}
+                    copy={
+                        <Interpolated
+                            id={gettext(
+                                'Support MDN with a $5 monthly subscription <learnMore/>.'
+                            )}
+                            learnMore={
+                                <a href={`/${locale}/payments/`}>
+                                    {gettext('Learn more')}
+                                </a>
+                            }
+                        />
+                    }
+                    cta={gettext('Subscribe')}
+                    url={`${
+                        window.mdn ? window.mdn.wikiSiteUrl : ''
+                    }/${locale}/profiles/${userData.username ||
+                        ''}/edit#subscription`}
+                    embargoDays={7}
+                />
+            );
+
+        default:
```

Comment:
How could this ever happen? The `bannerId` will never be anything that is NOT inside `BANNER_IDS`.
The only reason we could end up down here is if the `BANNER_IDS` array isn't the same as the list of blocks in this switch statement. 

It is, after all, kinda strange that we're actually maintaining two lists. One is the `BANNER_IDS` array. The other one is the list of case statements in the `switch` code. So that's not gorgeous but it'll be fine. 

Another way of structuring the code would be to put it like this:

```javascript
const BANNERS = {
  developer_needs: id => {
     return <Banner id={id} className="developer-needs" ... />
  },
  subscription_banner: id => {
      return <Banner id={id} className="subscription-banner" ... />
  }
}

for (const ([key, value]) of Object.entries(userData.waffle.flags)) {
  if (key in BANNERS && value && !isEmbargoed(key)) {
     return BANNERS[key](key);
  }
}
return null;
```

That way, the list of waffle flag keys is only listed 1 single time.